File: php/src/Model/SchedulePolicy.php
<?php

declare(strict_types=1);

namespace Fivexer\SDK\Model;

use Fivexer\SDK\Internal\Json;

/**
 * When a task may be offered.
 *
 * Unlike escalation and SLA there is no workspace default to inherit — these timestamps are
 * absolute epoch-milliseconds.
 */
final class SchedulePolicy
{
    public function __construct(
        /** Held out of matching until this moment; the task reads as 'scheduled' until then */
        private ?int $notBefore = null,
        /** The offer window closes here; an unserved task is parked or dropped */
        private ?int $notAfter = null,
        /** 'park' keeps a missed task for review, 'drop' discards it */
        private ?string $onMiss = null,
    ) {
    }

    public function notBefore(int $notBefore): self
    {
        $this->notBefore = $notBefore;
        return $this;
    }

    public function notAfter(int $notAfter): self
    {
        $this->notAfter = $notAfter;
        return $this;
    }

    public function onMiss(string $onMiss): self
    {
        $this->onMiss = $onMiss;
        return $this;
    }

    public function getNotBefore(): ?int
    {
        return $this->notBefore;
    }

    public function getNotAfter(): ?int
    {
        return $this->notAfter;
    }

    public function getOnMiss(): ?string
    {
        return $this->onMiss;
    }

    /** @return array<string, mixed> */
    public function toArray(): array
    {
        return Json::compact([
            'notBefore' => $this->notBefore,
            'notAfter' => $this->notAfter,
            'onMiss' => $this->onMiss,
        ]);
    }

    /** @param array<string, mixed> $data */
    public static function fromArray(array $data): self
    {
        return new self(
            isset($data['notBefore']) ? (int) $data['notBefore'] : null,
            isset($data['notAfter']) ? (int) $data['notAfter'] : null,
            isset($data['onMiss']) ? (string) $data['onMiss'] : null,
        );
    }
}

File: java/src/main/java/io/fivexer/sdk/model/SchedulePolicy.java
package io.fivexer.sdk.model;

/**
 * When a task may be offered.
 *
 * <p>Unlike escalation and SLA there is no workspace default to inherit — these timestamps are
 * absolute epoch-milliseconds.
 */
public class SchedulePolicy {

    private Long notBefore;
    private Long notAfter;
    private String onMiss;

    public SchedulePolicy() {}

    public Long getNotBefore() { return notBefore; }
    public Long getNotAfter() { return notAfter; }
    public String getOnMiss() { return onMiss; }

    /** Held out of matching until this moment; the task reads as {@code scheduled} until then. */
    public SchedulePolicy notBefore(long notBefore) { this.notBefore = notBefore; return this; }

    /** The offer window closes here; an unserved task is parked or dropped. */
    public SchedulePolicy notAfter(long notAfter) { this.notAfter = notAfter; return this; }

    /** {@code "park"} keeps a missed task for review, {@code "drop"} discards it. */
    public SchedulePolicy onMiss(String onMiss) { this.onMiss = onMiss; return this; }
}
